### PR TITLE
Sprint3 cleanup model associations

### PR DESCRIPTION
**Pull request by weedySeaDragon**
PT Story: **List member companies of a certain category**
https://www.pivotaltracker.com/story/show/135684057
This is a task needed so that  PR #88 can be accomplished.  See the comments in that PR for some of the reasoning behind this PR.

Now that we've learned more about the business, we need to ensure that the associations between the models can be traversed efficiently, and we need to delete old associations and/or tables not being used.

Changes proposed in this pull request:
1.  delete db table no longer used: 'business_categories_companies'
2.  Company:
   -  add these associations:
````
has_many :business_categories, through: :membership_applications	 
has_many :membership_applications
````
  - replace `.categories` with  `.business_categories`
  - update spec
3. CompanyHelper
    - rename the method from `list_categories` to `last_category_name` to reflect what it really does
    - add method that returns a string with all of the categories, separated by a space
    - update spec
4. BusinessCategory
    - added these associations:
````
  has_and_belongs_to_many :membership_applications
  has_many :companies, through: :membership_applications
````
5. use db_cleaner transaction settings for Rspec

Ready for review:
@patmbolger

### CODE REVIEW

File: spec/models/company_spec.rb [hunk 1]
```diff
@@ -32,58 +32,51 @@
...
+      m2 = create(:membership_application, user: User.find_by_email('emp2@happymutts.com'), num_categories: 0, status: 'Godkänd', company_number: '5562252998')
+      m2.business_categories << category
 
-      expect(company.categories.count).to eq 1
-      expect(company.categories.map { |c| c[:name] }).to contain_exactly('BusinessCategoryName')
+      m3 = create(:membership_application, user: User.find_by_email('emp3@happymutts.com'), num_categories: 0, status: 'Godkänd', company_number: '5562252998')
+      m3.business_categories << category
 
+      expect(Company.find_by_company_number('5562252998').business_categories.distinct.count).to eq 1
+      expect(Company.find_by_company_number('5562252998').business_categories.distinct.map { |c| c[:name] }).to contain_exactly('cat1')
```
Comment (patmbolger) [inline spec/models/company_spec.rb]: You're doing this the hard way :) ..... use instance vars (via `let` method) that span the suite and also negate the need to look up objects for `expect` tests.  That is, this test section could be written as:
```
describe 'categories = all employee categories' do

    let(:company) { create(:company, company_number: '5562252998') }

    let(:employee1) { create(:user) }
    let(:employee2) { create(:user) }
    let(:employee3) { create(:user) }

    let(:cat1) { create(:business_category, name: 'cat1') }
    let(:cat2) { create(:business_category, name: 'cat2') }
    let(:cat3) { create(:business_category, name: 'cat3') }

    let(:m1) do
      create(:membership_application, user: employee1,
             num_categories: 0, status: 'Godkänd',
             company_number: company.company_number)
    end
    let(:m2) do
      create(:membership_application, user: employee2,
             num_categories: 0, status: 'Godkänd',
             company_number: company.company_number)
    end
    let(:m3) do
      create(:membership_application, user: employee3,
             num_categories: 0, status: 'Godkänd',
             company_number: company.company_number)
    end

    before(:all) do
      Company.delete_all
      MembershipApplication.delete_all
      User.delete_all
    end

    it '3 employees, each with 1 unique category' do
      m1.business_categories << cat1
      m2.business_categories << cat2
      m3.business_categories << cat3

      expect(company.business_categories.count).to eq 3
      expect(company.business_categories.map(&:name))
        .to contain_exactly('cat1', 'cat2', 'cat3')
    end

    it '3 employees, each with the same category' do
      m1.business_categories << cat1
      m2.business_categories << cat1
      m3.business_categories << cat1

      expect(company.business_categories.distinct.count).to eq 1
      expect(company.business_categories.count).to eq 3
      expect(company.business_categories.distinct.map(&:name))
        .to contain_exactly('cat1')
    end
  end
```

File: app/helpers/companies_helper.rb [hunk 2]
```diff
@@ -4,9 +4,18 @@ def company_complete? company
     !company.nil? && !company.name.nil? && !company.name.empty? &&
       !company.region.nil? && !company.region.empty?
   end
+
+  def last_category_name company
+    if company.business_categories.any?
+      company.business_categories.last.name
+    end
+
+  end
```
Comment (patmbolger) [inline app/helpers/companies_helper.rb:11]: Would it be better to return "None" rather than `nil` since this is used in views?

**Review by patmbolger — COMMENTED** [review]
PR looks good ... please see comments for suggested changes.

File: spec/models/company_spec.rb (hunk 1 above)
Comment (weedySeaDragon) [reply spec/models/company_spec.rb]: Thank you.  :-) I was definitely struggling with the factories.

File: app/helpers/companies_helper.rb (hunk 2 above)
Comment (weedySeaDragon) [reply app/helpers/companies_helper.rb:11]: yes. will do.

**Review by thesuss — APPROVED** [review]

File: spec/helpers/companies_helper_spec.rb [hunk 3]
```diff
@@ -7,41 +7,53 @@
...
-      company = FactoryGirl.create(:company,
-                                   company_number: 5562252998)
-      expect(helper.list_categories(company)).not_to eq 'Träning'
+  describe 'companies' do
+
+    before(:all) do
+      Company.delete_all
+      MembershipApplication.delete_all
+      User.delete_all
+      company = FactoryGirl.create(:company, company_number: '5562252998')
```
[the remark this answers is not in the document]
Comment (patmbolger) [reply spec/helpers/companies_helper_spec.rb]: I suggest cleaning this up a bit like this:
```
require 'rails_helper'

RSpec.describe CompaniesHelper, type: :helper do
  let(:company) { create(:company) }

  describe '#company_complete?' do
    it 'returns false if company is nil' do
      expect(helper.company_complete?(nil)).to eq false
    end

    it 'returns false if company name is nil' do
      company.name = nil
      expect(helper.company_complete?(company)).to eq false
    end

    it 'returns false if company region is nil' do
      company.region = nil
      expect(helper.company_complete?(company)).to eq false
    end

    it 'returns false is company name or region is empty string' do
      company.name = ''
      expect(helper.company_complete?(company)).to eq false
    end

    it 'returns true if company name and region are non-empty string' do
      expect(helper.company_complete?(company)).to eq true
    end

  end

  describe 'companies' do
    let(:employee1) { create(:user) }
    let(:employee2) { create(:user) }
    let(:employee3) { create(:user) }

    let!(:ma1) { create(:membership_application, user: employee1,
                       status: 'Godkänd', category_name: 'cat1',
                       company_number: company.company_number) }
    let!(:ma2) { create(:membership_application, user: employee2,
                       status: 'Godkänd', category_name: 'cat2',
                       company_number: company.company_number) }
    let!(:ma3) { create(:membership_application, user: employee3,
                       status: 'Godkänd', category_name: 'cat3',
                       company_number: company.company_number) }


    before(:all) do
      Company.delete_all
      MembershipApplication.delete_all
      User.delete_all
    end

    it '#last_category_name' do
      expect(helper.last_category_name(company)).to eq 'cat3'
    end

    it '#list_categories' do
      expect(helper.list_categories(company)).to eq 'cat1 cat2 cat3'
      expect(helper.list_categories(company)).not_to include 'Träning'
    end
  end
end
```

**Review by patmbolger — COMMENTED** [review]
I have one more set of suggested changes to companies_helper_spec.rb

**Pull request merged**